File: maven/Pkcs10Sample.java
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.Security;

import org.bouncycastle.asn1.x509.X509Name;
import org.bouncycastle.jce.PKCS10CertificationRequest;
import org.bouncycastle.jce.provider.BouncyCastleProvider;

public class Pkcs10Sample {
    public static void main(String[] args) throws Exception {
        Security.addProvider(new BouncyCastleProvider());

        KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA", "BC");
        generator.initialize(2048);
        KeyPair keyPair = generator.generateKeyPair();

        X509Name subjectDn = new X509Name("CN=localhost");
        PKCS10CertificationRequest req = new PKCS10CertificationRequest(
                "SHA1withRSA", subjectDn, keyPair.getPublic(), null,
                keyPair.getPrivate());

        byte[] bin = req.getEncoded();
        System.out.println("Generated: " + bin.length + "[bytes]");
    }
}

File: jjug-jruby-2011/maven/pkcs10sample.rb
require 'java'
require 'rubygems'
require 'maven/bouncycastle/bcprov-jdk15'

java_import 'java.security.KeyPair'
java_import 'java.security.KeyPairGenerator'
java_import 'java.security.Security'

java_import 'org.bouncycastle.asn1.x509.X509Name'
java_import 'org.bouncycastle.jce.PKCS10CertificationRequest'
java_import 'org.bouncycastle.jce.provider.BouncyCastleProvider'

Security.add_provider(BouncyCastleProvider.new)

generator = KeyPairGenerator.get_instance("RSA", "BC")
# method 'initialize' conflicts with the method name in Ruby world.
# invoking it by java_send with specifying method signature.
generator.java_send :initialize, [Java::int], 2048
key_pair = generator.generate_key_pair

subject = X509Name.new("CN=localhost")
req = PKCS10CertificationRequest.new("SHA1withRSA",
  subject, key_pair.public, nil, key_pair.private)

bin = String.from_java_bytes(req.encoded)
puts "Generated: #{bin.bytesize}[bytes]"
